### ISSUE TITLE
DataController#extendSelection should not move selection to the inside of an object

### ISSUE THREAD

**Issue by szymonkups** [ckeditor/ckeditor5#3904]
Extracted from ckeditor/ckeditor5#3893 @Reinmar's comment:

General rule of thumb – selection should not be moved to the inside of an object. E.g.:

`<obj></obj><p>^x</p>` + backspace => the selection must not be moved to `<obj>^</obj>`

This means that extending the selection to the left in the case mentioned above should not put the focus inside the object. The focus must be placed before the object.

* `<obj></obj><p>^x</p>` + extend left -> `[<obj></obj><p>]x</p>`,
* `<obj></obj>[<obj></obj>]` + extend left -> `[<obj></obj><obj></obj>]`,

So in other words, the selection outputed from the `extendSelection` should not cross object boundaries.
